watchdog.go

```diff
@@ -26,239 +76,316 @@ var (
+	}
 
-	// Policy sets the firing policy of this watchdog.
-	Policy Policy
+	_watchdog.scope = UtilizationHeap
+	_watchdog.closing = make(chan struct{})
 
-	// NotifyFired, if non-nil, will be called when the policy has fired,
-	// prior to calling GC, even if GC is disabled.
-	NotifyFired func()
+	var gcTriggered = make(chan struct{}, 16)
+	setupGCSentinel(gcTriggered)
 
-	// NotifyOnly, if true, will cause the watchdog to only notify via the
-	// callbacks, without triggering actual GC.
-	NotifyOnly bool
+	_watchdog.wg.Add(1)
+	go func() {
+		defer _watchdog.wg.Done()
 
-	// Logger is the logger to use. If nil, it will default to a logger that
-	// proxies to a standard logger using the "[watchdog]" prefix.
-	Logger Logger
-}
+		// get the initial effective GOGC; guess it's 100 (default), and restore
+		// it to whatever it actually was. This works because SetGCPercent
+		// returns the previous value.
+		originalGOGC := debug.SetGCPercent(debug.SetGCPercent(100))
+		currGOGC := originalGOGC
 
-// Memory starts the singleton memory watchdog with the provided configuration.
-func Memory(config MemConfig) (err error, stop func()) {
-	if !atomic.CompareAndSwapInt64(&watchdog.state, stateUnstarted, stateStarted) {
-		return ErrAlreadyStarted, nil
-	}
+		var memstats runtime.MemStats
+		for {
+			select {
+			case <-gcTriggered:
+				NotifyFired()
 
-	if config.Logger == nil {
-		config.Logger = &stdlog{log: log.New(log.Writer(), "[watchdog] ", log.LstdFlags|log.Lmsgprefix)}
-	}
-
-	// if the user didn't provide a limit, get the total memory.
-	if config.Limit == 0 {
-		var mem gosigar.Mem
-		if err := mem.Get(); err != nil {
-			return fmt.Errorf("failed to get system memory limit via SIGAR: %w", err), nil
-		}
-		config.Limit = mem.Total
-	}
+			case <-_watchdog.closing:
+				return
+			}
 
-	watchdog.config = config
-	watchdog.closing = make(chan struct{})
+			// recompute the next trigger.
+			memstatsFn(&memstats)
 
-	watchdog.wg.Add(1)
-	go watch()
+			// heapMarked is the amount of heap that was marked as live by GC.
+			// it is inferred from our current GOGC and the new target picked.
+			heapMarked := uint64(float64(memstats.NextGC) / (1 + float64(currGOGC)/100))
+			if heapMarked == 0 {
+				// this shouldn't happen, but just in case; avoiding a div by 0.
+				continue
+			}
 
-	return nil, stopMemory
-}
+			// evaluate the policy.
+			next, immediate := policy.Evaluate(UtilizationHeap, memstats.HeapAlloc)
+
+			if immediate {
+				// trigger a forced GC; because we're not making the finalizer
+				// skip sending to the trigger channel, we will get fired again.
+				// at this stage, the program is under significant pressure, and
+				// given that Go GC is not STW for the largest part, the worse
+				// thing that could happen from infinitely GC'ing is that the
+				// program will run in a degrated state for longer, possibly
+				// long enough for an operator to intervene.
```

Comment:
It's worse. If we're over-subscribed but not actually using all the allocated memory, we'll run GC in a loop even before we run out of memory. We should really work hard to avoid this.